k2/csrc/determinize.cc

```diff
@@ -0,0 +1,230 @@
+
+     The forward_backward_weight is the weight on the best path through the
+     output determinized FSA that will include this DetState.  It will determine
+     the order of expansion of DetStates and also whether the states are
+     expanded at all (if the pruning beam `beam` is finite).
+     forward_backward_weight is the sum of the forward weight of the base state,
+     plus (the greatest over the DetStateElements, of its `weight` element,
+     plus the backward weight in the input FSA of the state that corresponds
+     to it).
+
+
+     worked outobtained from
+
+   */
+  void Normalize(std::vector<int32_t> *leftover_arcs);
+};
+
+
+void DetState::Normalize(std::vector<int32_t> *input_arcs) {
+
+}
+
+
+class DetStateMap {
+ public:
+  /*
+    Outputs the output state-id corresponding to a specific DetState structure.
+    This does not store any pointers to the DetState or its contents, so
+    you can delete the DetState without affecting this object's ability to map
+    an equivalent DetState to the same state-id.
+
+       @param [in] a  The DetState that we're looking up
+       @param [out] state_id  The state-index in the output FSA
+                      corresponding to this DetState (will
+                      be freshly allocated if an equivalent of
+                      this DetState did not already exist.
+        @return  Returns true if this was a NEWLY CREATED state,
+              false otherwise.
+   */
+  bool GetOutputState(const DetState &a, int32_t *state_id) {
+    std::pair<uint64_t, uint64_t> compact;
+    DetStateToCompact(a, &compact);
+    auto p = map_.insert({compact, cur_output_state));
+    bool inserted = p.second;
+    if (inserted) {
+      *state_id = cur_output_state_++;
+      return true;
+    } else {
+      *state_id = p.first->second;
+      return false;
+    }
+  }
+
+  int32_t size() const { return cur_output_state_; }
+
+ private:
+
+  int32_t cur_output_state_ { 0 };
+  std::unordered_map<std::pair<uint64_t, uint64_t>, int32_t, DetStateVectorHasher> map_;
+
+  /* Turns DetState into a compact form of 128 bits.  Technically there
+     could be collisions, which would be fatal for the algorithm, but this
+     is one of those lifetime-of-the-universe type of things (kind of like
+     the theoretical potential for git hash collision) that we ignore. */
+  void DetStateToCompact(const DetState &d,
+                         std::pair<uint64_t, uint64_t> *vec) {
+    assert(d.normalized);
+
+    uint64_t a = d.base_state + 17489 * d.seq_len,
+        b = d.base_state * 103979  + d.seq_len;
+
+    // We choose an arbitrary DetStateElement (the first one in the list) to
+    // read the symbol sequence from; the symbol sequence will be the same no
+    // matter which element we choose to trace back.
+    DetStateElement *elem = d.head;
+    int32_t seq_len = d.seq_len;
+    for (int32_t i = 0; i < seq_len; i++) {
+      a = elem->symbol + 102299 * a;
```

Comment:
Is there an error inside the `for` loop?

`elem` is never updated.

----

It should be `++i`, not `i++`.